test/jasmine/tests/cartesian_interact_test.js

```diff
@@ -2547,3 +2547,123 @@ describe('Cartesian plots with css transforms', function() {
         });
     });
 });
+
+describe('Cartesian taces with zindex', function() {
+    var gd;
+
+    beforeEach(function() {
+        gd = createGraphDiv();
+    });
+
+    afterEach(destroyGraphDiv);
+
+    var data0 = [
+        {x: [1, 2], y: [1, 1], type: 'scatter', zindex: 10},
+        {x: [1, 2], y: [1, 2], type: 'scatter'},
+        {x: [1, 2], y: [1, 3], type: 'scatter', zindex: 5}
+    ];
+
+    var data1 = [
+        {x: [1, 2], y: [1, 1], type: 'scatter'},
+        {x: [1, 2], y: [1, 2], type: 'scatter', zindex: -5},
+        {x: [1, 2], y: [1, 3], type: 'scatter', zindex: 10},
+    ];
+
+    function fig(data) {
+        return {
+            data: data,
+            layout: {
+                width: 400, height: 400
```

Comment:
For bar traces let's test this using `barmode: 'overlay'` as well as `barmode: 'stack'`.